File: src/main/java/net/kenro/ji/jin/purescript/file/PSFile.java
package net.kenro.ji.jin.purescript.file;

import com.intellij.extapi.psi.PsiFileBase;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.psi.FileViewProvider;
import net.kenro.ji.jin.purescript.PSLanguage;
import net.kenro.ji.jin.purescript.psi.impl.PSProgramImpl;
import net.kenro.ji.jin.purescript.psi.impl.PSValueDeclarationImpl;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.Map;

public class PSFile extends PsiFileBase {
    public PSFile(@NotNull final FileViewProvider viewProvider) {
        super(viewProvider, PSLanguage.INSTANCE);
    }

    @NotNull
    @Override
    public FileType getFileType() {
        return PSFileType.INSTANCE;
    }

    @Override
    public String toString() {
        return "Purescript File";
    }

    @Override
    public Icon getIcon(final int flags) {
        return super.getIcon(flags);
    }

    public Map<String, PSValueDeclarationImpl> getTopLevelValueDeclarations() {
        return this
            .getProgram()
            .getModule()
            .getTopLevelValueDeclarations()
        ;
    }

    private PSProgramImpl getProgram() {
        return this.findChildByClass(PSProgramImpl.class);
    }
}

File: src/main/kotlin/org/purescript/file/PSFile.kt
package org.purescript.file

import com.intellij.extapi.psi.PsiFileBase
import com.intellij.openapi.fileTypes.FileType
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiFile
import com.intellij.psi.StubBuilder
import com.intellij.psi.stubs.DefaultStubBuilder
import com.intellij.psi.stubs.PsiFileStubImpl
import com.intellij.psi.tree.IStubFileElementType
import org.purescript.PSLanguage
import org.purescript.module.Module
import java.nio.file.Path
import java.nio.file.Paths

class PSFile(viewProvider: FileViewProvider) :
    PsiFileBase(viewProvider, PSLanguage) {
    class Stub(file: PSFile) : PsiFileStubImpl<PSFile>(file) {
        override fun getType() = Type
    }

    object Type : IStubFileElementType<Stub>("PSFile", PSLanguage) {
        override fun getStubVersion(): Int = 14
        override fun getBuilder(): StubBuilder = object : DefaultStubBuilder() {
            override fun createStubForFile(file: PsiFile): Stub = Stub(file as PSFile)
        }
    }
        override fun getFileType(): FileType = PSFileType
        override fun toString(): String = "Purescript File"

        /**
         * @return the [Module] that this file contains,
         * or null if the module couldn't be parsed
         */
        val module: Module?
            get() = findChildByClass(Module::class.java)

    fun suggestModuleName(): String? {
        val fileName = name.removeSuffix(".purs")
        val path = parent?.virtualFile?.path ?: return null
        val directoryPath = Paths.get(path)
        val relativePath = try {
            project
                .basePath
                ?.let { Paths.get(it).toAbsolutePath() }
                ?.relativize(directoryPath.toAbsolutePath())
                ?: directoryPath
        } catch (ignore: IllegalArgumentException) {
            directoryPath
        }
        return relativePath
            .reversed<Path?>()
            .takeWhile<Path?> { "$it" != "src" && "$it" != "test" }
            .filter<Path?> { "$it".first().isUpperCase() }
            .reversed<Path?>()
            .joinToString<Path?>(".")
            .let<String, String> {
                "$it.$fileName"
            }
            .removePrefix(".")
    }

}
